stack-graphs/src/filtered_stack_graph.rs

```diff
@@ -0,0 +1,386 @@
+    pub fn reconstruct(&self) -> crate::graph::StackGraph {
+        todo!()
+    }
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+#[serde(transparent)]
+pub struct Files {
+    data: Vec<String>,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+#[serde(transparent)]
+pub struct Nodes {
+    data: Vec<Node>,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+#[serde(tag = "type")]
+pub enum Node {
+    #[serde(rename = "drop_scopes")]
+    DropScopes {
+        id: NodeID,
+        source_info: Option<SourceInfo>,
+        debug_info: Option<DebugInfo>,
+    },
+
+    #[serde(rename = "jump_to")]
+    JumpTo {
+        id: NodeID,
+        source_info: Option<SourceInfo>,
+        debug_info: Option<DebugInfo>,
+    },
+
+    #[serde(rename = "pop_scoped_symbol")]
+    PopScopedSymbol {
+        id: NodeID,
+        symbol: String,
+        is_definition: bool,
+        source_info: Option<SourceInfo>,
+        debug_info: Option<DebugInfo>,
+    },
+
+    #[serde(rename = "pop_symbol")]
+    PopSymbol {
+        id: NodeID,
+        symbol: String,
+        is_definition: bool,
+        source_info: Option<SourceInfo>,
+        debug_info: Option<DebugInfo>,
+    },
+
+    #[serde(rename = "push_scoped_symbol")]
+    PushScopedSymbol {
+        id: NodeID,
+        symbol: String,
+        scope: NodeID,
+        is_reference: bool,
+        source_info: Option<SourceInfo>,
+        debug_info: Option<DebugInfo>,
+    },
+
+    #[serde(rename = "push_symbol")]
+    PushSymbol {
+        id: NodeID,
+        symbol: String,
+        is_reference: bool,
+        source_info: Option<SourceInfo>,
+        debug_info: Option<DebugInfo>,
+    },
+
+    #[serde(rename = "root")]
+    Root {
+        id: NodeID,
+        source_info: Option<SourceInfo>,
+        debug_info: Option<DebugInfo>,
+    },
+
+    #[serde(rename = "scope")]
+    Scope {
+        id: NodeID,
+        is_exported: bool,
+        source_info: Option<SourceInfo>,
+        debug_info: Option<DebugInfo>,
+    },
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct SourceInfo {
+    span: lsp_positions::Span,
+    syntax_type: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+#[serde(transparent)]
+pub struct DebugInfo {
+    data: Vec<DebugEntry>,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct DebugEntry {
+    key: String,
+    value: String,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct NodeID {
+    file: Option<String>,
+    local_id: u32,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+#[serde(transparent)]
+pub struct Edges {
+    data: Vec<Edge>,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct Edge {
+    source: NodeID,
+    sink: NodeID,
+    precedence: i32,
+}
+
+impl crate::graph::StackGraph {
+    pub fn apply_filter<'a>(&self, filter: &'a dyn Filter) -> StackGraph {
```

Comment:
I would perhaps define this the other way around:

```rust
impl StackGraph {
   fn from_graph(graph: &StackGraph, filter: &'a dyn Filter) -> Self {
      // ...
   }
}
```

In this direction I would maybe call it `to_filtered_graph`, which aligns with common Rust method naming. We could have both directions, for more flexibility for the user, I guess.

Reply:
yeah that makes sense. 

i would also like to stick a discussion about the public API in here, i was hoping the public API of `crate::graph::StackGraph` might look like:

- `crate::graph::StackGraph::to_json(&self) -> serde_json::Value`
- `crate::graph::StackGraph::to_json_with_filter(&self, filter: _) -> serde_json::Value`

orthogonally, we do not have to limit this to `to_json` at all, because we have the ability to serialize to any format of choice:

- `crate::graph::StackGraph::to_serializeable(&self) -> filtered_stack_graph::StackGraph` (which can then be passed to `serde_yaml::to_value` or `serde_json::to_value`), which might, at the call-site, look like this: `serde_json::to_value(&graph.to_serializable())`
- likewise with a `filter`